### ISSUE TITLE
Postgre: Usage feedback

### ISSUE THREAD

**Issue by bjg2** [beekeeper-studio/beekeeper-studio#173]
Hi,

first of all, thanks for the effort!

I agree the issue exists and that app like this is very much needed, that's why installed it right away and tried it out. I played with it for a few minutes, and because you are so kind to create this for everyone I spent few minutes to give you feedback on what I consider problems that blocks me for using it. You don't have to fix those, I just wanted to notify you what people might find problems when starting beekeeper.

* Don't include all schemas and tables to the sidebar. At least hide bunch of them by default and don't wait on their load on open. It is just confusing. I don't need information_schema, pg_catalog, etc in sidebar nor I want to wait for load of those info. Also, maybe split tables and views, maybe add functions, etc.
* Table code complete does not work for non-public schema. I have a schema render and it does not give auto-complete suggestions for tables in it.
![image](https://user-images.githubusercontent.com/6639273/83394657-3c2d3000-a3f9-11ea-894a-874bf9ba7352.png)
* I expect to be able to interact with tables in the sidebar. Tried to double click it or drag-n-drop it to the query but nothing happens.
* I expect F5 to run the query (ctrl + enter is kind of weird)
* I expect that I can select only a part of the query to run. Creating new query and copying to it is slow.
* I expect to be able to select multiple lines and comment them all at the same time
* Queries that do return info but not rows are not written out in any way. For example:
![image](https://user-images.githubusercontent.com/6639273/83395005-d68d7380-a3f9-11ea-9a4c-e8d141215158.png)
* By far the hardest one - it is very important for me to be able to see the query plan... I need to be able to see some graphical representation of it...

I'm writting all of this because I really like look and feel of the beekeper, it is really cool. Anyway, thanks for everything!

**Comment by phillpafford** [comment]
I would agree with everything, but I do want to see all schemas in the sidebar, so maybe an option to load all/public when first opening the app?

Also I noticed it takes a while to load all the schemas/tables/etc... on initial start. working on large projects that can have hundreds, if not thousands of database tables, the app loads very slowly before I can even query a table. Possibly cache the initial load so sequential app start times would be improved?

**Comment by rathboma** [comment]
Awesome feedback! Thanks so much.

Many of these we know about and we're working to solve.

The schema ux is definitely wonky right now, but we're slowly moving forwards. I think a setting for the connection would make sense for sure to optionally limit schemas --- many folks use them like databases, so we can't hide them all the time.

Thanks for filing! I think you can see our vision, we're working furiously to fill in the gaps :-)

**Comment by GammaGames** [comment]
> I don't need information_schema, pg_catalog, etc in sidebar

I agree. I am okay with seeing it on first connection, but the ability to right click and hide a table or view (as well as right click and show hidden) would be a nice

> maybe split tables and views

I don't agree with this, part of the reason I like beekeeper so much is because of how simple it is. No nested trees to dig into to see the tables, I've seen other management tools that need you to go into database > "Schemas" > schema > "Tables" > table.  


> maybe add functions

If functions, roles, sequences, etc are added I think the could be put them in additional panes that could be collapsed and expanded would be clean and help keep the interface uncluttered. Extensions (postgis for example) and other information could be put in the activity bar with plug-ins too

**Comment by ricobl** [comment]
I've been looking for a modern multi-db client and I see a lot of potential for Beekeeper. As mentioned, the UI seems really cool and I'll keep an eye on the project.

Regarding postgres, I also noticed that connecting can be really slow and I believe it's related to displaying the schema on the sidebar.

I've enabled query logging in my local machine and TablePlus uses 7 queries when connecting while Beekeeper uses around 190 on the same DB with only 5 tables.

So postponing some of these might improve performance, specially when tunneling. When using TablePlus, clicking on a table will display its data and shows a button to display the structure, I believe that works well for most people as the schema is usually pretty static compared to data.

**Comment by rathboma** [comment]
@ricobl so currently Beekeeper performs:
- 1 query to get all tables
- 1 query PER table to get columns

So if you have 150 tables, it makes 151 queries.

This isn't ideal. We used to only load in columns when you expanded the table in the sidebar, but then we couldn't effectively use column names for auto-complete in the query editor. So this was a quick fix to get that working.

What we SHOULD do is:
- 1 query to get all tables
- 1 query to get all columns
- match columns to tables

That would make start-up times significantly faster

**Comment by mariusa** [comment]
After connecting to Heroku postgres, Beekeeper shows > 100 databases under "Connection" dropdown. I already specified the database when creating the connection. Is reading all dbs necessary? One might have a db dropdown when creating the connection, but I don't see why it's needed afterwards.

**Comment by rathboma** [comment]
@mariusa Yes for Heroku I can see how that is not ideal. I think we should have the option to 'hide' other databases on the connection.

I don't think this should be the default though -- many folks use multiple databases in their production set-up so we don't want to obfuscate that.

**Comment by phillpafford** [comment]
> @ricobl so currently Beekeeper performs:
> 
> * 1 query to get all tables
> * 1 query PER table to get columns
> 
> So if you have 150 tables, it makes 151 queries. It looks at ALL tables, regardless of schema.
> 
> This isn't ideal. We used to only load in columns when you expanded the table in the sidebar, but then we couldn't effectively use column names for auto-complete in the query editor. So this was a quick fix to get that working.
> 
> What we SHOULD do is:
> 
> * 1 query to get all tables
> * 1 query to get all columns
> * match columns to tables
> 
> That would make start-up times significantly faster

Any update on this change? this is one of the main reasons I can not use as my main db tool

**Comment by rathboma** [comment]
Hey @phillpafford this is in the latest version and working great. It does 2 queries (well 4 due to materialized views)

Also a big release coming later this week with:
- table editing
- Foreign keys linking to the foreign key table (one click go-go)
- sql formatting

**Comment by phillpafford** [comment]
Not sure if I should just open another issue but I feel this one has some great feedback. I'm one of those users that is still using PGAdmin 3 LTS by BigSQL (Can no longer find). I feel Beekeeper is the best solution to replace it but still missing some things. I think I've seen other open issues addressing some of them but I will add them again just to bring more attention to them

- DDL of all things 
![image](https://user-images.githubusercontent.com/655856/96647323-67a6ad80-12fb-11eb-8257-bb290377f483.png)

in PGAdmin I could drill down into the table, triggers, functions, etc.. and see the raw SQL

- Visual Query Planner 
![image](https://user-images.githubusercontent.com/655856/96647608-fa474c80-12fb-11eb-9397-8c5a885594d9.png)

the visual query planner is so helpful in understanding what is happening, if indexes are used, etc... 

- Autocomplete for all schemas, just not public

When referencing a different schema in the SQL does not auto complete

- EXPLAIN ANALYZE 
![image](https://user-images.githubusercontent.com/655856/96648248-10a1d800-12fd-11eb-9ea0-d76ef218704f.png)

Should be an open in the Run Button

- Export Settings: I have several machines and several database connections, would be great if I could export/import a config with all my username / connection string / etc... 

- Scratch Pad: an area where I could just save random text/results/etc...

Thanks to all your hard work on Beekeeper I'm very close to being able to not use PGAdmin 3 much appreciated!

**Comment by rathboma** [comment]
Hey folks. I'm closing this issue as I'm tracking individual ideas as separate issues.

I am in no way ignoring these pieces of feedback. In fact after adding Oracle support I intend to dedicate a lot of time to Postgresql users. :heart:
